src/observer.js

```diff
@@ -0,0 +1,81 @@
+import { parseRootMargin, shallowCompare } from './utils';
+
+export const observerElementsMap = new Map();
+
+export function getPooled(options = {}) {
+    const root = options.root || null;
+    const rootMargin = parseRootMargin(options.rootMargin);
+    const threshold = Array.isArray(options.threshold)
+        ? options.threshold
+        : [options.threshold != null ? options.threshold : 0];
+    const observers = observerElementsMap.keys();
+    let observer;
+    while ((observer = observers.next().value)) {
+        const unmatched =
+            root !== observer.root ||
+            rootMargin !== observer.rootMargin ||
+            shallowCompare(threshold, observer.thresholds);
+
+        if (!unmatched) {
+            return observer;
+        }
+    }
+    return null;
+}
+
+export function findObserverElement(observer, entry) {
+    const elements = observerElementsMap.get(observer);
+    if (elements) {
+        const values = elements.values();
+        let element;
+        while ((element = values.next().value)) {
+            if (element.target === entry.target) {
+                return element;
+            }
+        }
+    }
+    return null;
+}
+
+/**
+ * The Intersection Observer API callback that is called whenever one element,
+ * called the target, intersects either the device viewport or a specified element.
+ * Also will get caled whenever the visibility of the target element changes and
+ * crosses desired amounts of intersection with the root.
+ * @param {array} changes
+ * @param {IntersectionObserver} observer
+ */
+export function callback(changes, observer) {
+    for (let i = 0; i < changes.length; i++) {
+        const element = findObserverElement(observer, changes[i]);
+        if (element) {
+            element.handleChange(changes[i]);
+        }
+    }
+}
+
+export function createObserver(options) {
+    return getPooled(options) || new IntersectionObserver(callback, options);
+}
+
+export function observeElement(element) {
```

Comment:
This was slightly smaller before, any reason for the change?

Reply:
Minified comparison:

```js
// before 134 bytes (ignoring additional class overhead):
var e=new Map;function observe(n){var v=void 0;if(e.has(n.v)){v=e.get(n.v)}else{v=new Set;e.set(n.v,v)}v.add(n);n.v.observe(n.target)}
// after 116 bytes:
var e=new Map;function observeElement(n){if(!e.has(n.o)){e.set(n.o,new Set)}e.get(n.o).add(n);n.o.observe(n.target)}
```

In addition I think the new version is a lot easier to understand, oftentimes with `let variable` and then filling it afterwards inside `if else` makes it harder to read what is happening.